post: git rebase interactive (#111)
cloudx-labs/posts: posts/navarroaxel/18.git-rebase-interactive.md

---
title: Rewriting the history with Git rebase
published: true
description: How to squash commits and keep your branch up-to-date using git rebase 
tags: 'git, bash, productivity, tooling'
cover_image: ./assets/git-rebase.jpg
id: 1949937
---

> As when we work in a new feature or bug fix it is recommended to keep our branch up-to-date with the target branch (`main`, `development`, etc.), we can use the [`git rebase`](https://git-scm.com/docs/git-rebase) command instead of `git merge` in order to keep our Git history cleaner.

Reviewer: ```suggestion
As when we work in a new feature or bug fix it is recommended to keep our branch up-to-date with the target branch (`main`, `development`, etc.), we can use the [`git rebase`](https://git-scm.com/docs/git-rebase) command instead of `git merge` in order to keep our Git history clean.
```

> The `git rebase` command reapplies our commits on top of another branch, rewriting the Git history. This is not recommended on shared branches (like `main`) because it causes divergence (and therefore conflict solving) and panic on your teammates 😬 since it is sometimes hard to resolve.

Reviewer: ```suggestion
The `git rebase` command reapplies our commits on top of the target branch, rewriting the Git history. This is not recommended on shared branches (like `main`) because it causes divergence (and therefore the need of conflict solving) and panic on your teammates 😬 since it is sometimes hard to resolve.
```

You can fix the problem, and then run

  git rebase --continue
```

Sad time when you need to fix the tests 😢. Before continuing the rebase you will need to commit these changes creating a new commit; or integrating there in the latest one using the classic [`git commit --amend`](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---amend).

```bash
git add .
git commit --amend
git rebase --continue
```

The `reword` command that we put into the `git-rebase-todo` file will open your default editor (e.g. nano) with the current commit message for `76d14ac`. Edit it and save to continue.

💡 The `git rebase` reapplies commits, you can do this without needing a target branch (`main`): just indicate how many commits you want to reapply for the current branch (`HEAD`). For our example of 3 latest commits:

```bash
git rebase --interactive HEAD~3
```

🧠 Remember that `HEAD` is a pointer to either your current branch or your current commit when you're in a **detached HEAD** (meaning you are located in a commit not associated with any branch).

And, this is the usage of `git rebase --interactive` 🎉. Let's continue with the `git rebase --exec`.

> Do you prefer `rebase` or `merge` to keep your branch up to date? 👇 Let me know in the comments.

Reviewer: ```suggestion
Do you either prefer `rebase` or `merge` to keep your branch up to date? 👇 Let me know in the comments.
```
